### ISSUE TITLE
Potential TOCTOU Vulnerability in getBundleManifest() File Handling

### ISSUE THREAD

**Issue by naveensrinivasan** [defenseunicorns/uds-cli#440]
A potential Time-Of-Check to Time-Of-Use (TOCTOU) vulnerability has been identified in the `getBundleManifest` function within the tarballBundleProvider, specifically within the `tarball.go` file. This vulnerability arises because the files `indexPath` and `manifestPath` are marked for deletion immediately after their creation and before the processing of their contents, creating an opportunity for an attacker to replace these files with malicious versions between their verification and use.

https://research.nccgroup.com/wp-content/uploads/2021/09/TOCTOU_whitepaper.pdf
### Attack Scenario
Initial State: The application creates temporary files (indexPath and manifestPath) to store data temporarily. These files are marked for deletion (using defer os.Remove(filePath)) after their contents are processed.

Exploitation Window: Between the creation (and subsequent check) of these files and their actual usage by the application, there's a brief time window. This is due to the deferred deletion, which does not occur until the function returns. During this window, the files are vulnerable to tampering.

### Attack Execution:

An attacker, with access to the file system where the application is running, monitors the creation of these temporary files.
Once the application creates indexPath or manifestPath, the attacker quickly replaces one or both of these files with malicious versions. The content of these malicious files could be crafted to exploit vulnerabilities in the application's file processing logic or to execute arbitrary code.
Impact:

The application proceeds to use the tampered files, under the assumption that they have not been altered since their creation and initial check.
Depending on the nature of the tampered content and the application's functionality, this could lead to various adverse outcomes, such as data corruption, unauthorized data disclosure, execution of arbitrary code, or complete system compromise.

```go
indexPath := filepath.Join(tp.dst, "index.json")
defer os.Remove(indexPath)
...
manifestPath := filepath.Join(tp.dst, manifestRelativePath)
defer os.Remove(manifestPath)
```

The potential exploitation of this vulnerability could allow for the compromise of the application's integrity or enable the execution of arbitrary code if an attacker has the capability to manipulate the file system.

Mitigation strategies may include:
1. Assessing the necessity of deleting these files immediately after processing. If immediate deletion is not critical, it might be safer to remove the files only after their contents have been securely processed.
2. Implementing integrity checks for the files just before they are used. This could be achieved by verifying the files against pre-determined good signatures or checksums.
3. Employing atomic operations for file handling to minimize the risk associated with this vulnerability.

To mitigate the risk, a checksum verification step can be introduced before the files are processed. Here is an example in Go that demonstrates how to calculate and verify the SHA-256 checksum of a file, ensuring its integrity has not been compromised between the time of check and the time of use:

```go
import (
    "crypto/sha256"
    "encoding/hex"
    "io/ioutil"
)

// calculateFileChecksum calculates the SHA-256 checksum of a file.
func calculateFileChecksum(filePath string) (string, error) {
    data, err := ioutil.ReadFile(filePath)
    if err != nil {
        return "", err
    }
    checksum := sha256.Sum256(data)
    return hex.EncodeToString(checksum[:]), nil
}

// verifyFileChecksum compares the calculated checksum of a file against an expected checksum.
func verifyFileChecksum(filePath, expectedChecksum string) (bool, error) {
    calculatedChecksum, err := calculateFileChecksum(filePath)
    if err != nil {
        return false, err
    }
    return calculatedChecksum == expectedChecksum, nil
}
```

### Examples of TOCTOU in the Wild
**Race Condition in Unix SUID Programs**: One of the classic examples of TOCTOU vulnerabilities is found in Unix-like systems with Set-User-ID (SUID) programs. Attackers exploit the time window between when a SUID program checks a file's attributes (like its owner or permissions) and when it performs operations on the file, potentially allowing the attacker to substitute the file with a symlink to a critical system file, leading to privilege escalation.

**Symbolic Link (Symlink) Attacks**: Symlink attacks are a common manifestation of TOCTOU vulnerabilities where an attacker creates a symbolic link to a critical file or resource that a software is expected to create or modify. The software checks for the resource's state (non-existent or safe to write), but before it performs the operation, the attacker's symlink is placed, leading the software to unintentionally modify or disclose sensitive information stored in a different file.

**Web Application File Upload Vulnerabilities**: In web applications, a TOCTOU vulnerability might occur in the handling of file uploads. An application might check if an uploaded file is of a safe type (e.g., a JPEG image) but fail to securely handle the file afterward. An attacker could exploit this by initially uploading a safe file, then changing it to a malicious script between the check and the file's execution or storage, leading to server compromise or data breaches.

Incorporating such a verification step ensures the integrity of `indexPath` and `manifestPath` before they are utilized, thereby enhancing the security posture against potential TOCTOU attack scenarios.

**Comment by UncleGedd** [comment]
Great catch! Will prioritize and fix the vulnerability. Appreciate the checksum example, that looks like the way to go
